librz/util/subprocess.c

```diff
@@ -927,53 +985,67 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
 	int stdin_pipe[2] = { -1, -1 };
 	int stdout_pipe[2] = { -1, -1 };
 	int stderr_pipe[2] = { -1, -1 };
-	if (opt->stdin_pipe == RZ_SUBPROCESS_PIPE_CREATE) {
-		if (rz_sys_pipe(stdin_pipe, true) == -1) {
-			perror("pipe");
-			goto error;
-		}
-		proc->stdin_fd = stdin_pipe[1];
+
+	/* We won't be creating any new pipes in case of forkpty */
+	if (opt->fork_mode != RZ_SUBPROCESS_FORKPTY && !init_pipes(proc, opt, stdin_pipe, stdout_pipe, stderr_pipe)) {
+		goto error;
 	}
 
-	if (opt->stdout_pipe == RZ_SUBPROCESS_PIPE_CREATE) {
-		if (rz_sys_pipe(stdout_pipe, true) == -1) {
-			perror("pipe");
-			goto error;
+	// Let's create the environment for the child in the parent, with malloc,
+	// because we can't use functions that lock after fork
+	child_env = create_child_env(opt->envvars, opt->envvals, opt->env_size);
+
+	int master_fd = -1;
+	void *term_params = NULL;
+
+	if (opt->fork_mode == RZ_SUBPROCESS_FORKPTY) {
+#if HAVE_FORKPTY && HAVE_OPENPTY && HAVE_LOGIN_TTY
+		struct termios t;
+		/* Needed to avoid reading back the writes again from the TTY */
+		tcgetattr(STDIN_FILENO, &t);
+		cfmakeraw(&t);
+		term_params = &t;
+#endif
+		if (opt->pty) {
+			pty = opt->pty;
+		} else {
+			pty = rz_subprocess_openpty(NULL, term_params, NULL);
 		}
-		if (fcntl(stdout_pipe[0], F_SETFL, O_NONBLOCK) < 0) {
-			perror("fcntl");
+
+		if (pty) {
+			proc->master_fd = master_fd = pty->master_fd;
+			if (tcsetattr(pty->slave_fd, TCSANOW, term_params) == -1) {
+				perror("tcsetattr");
```

Comment:
what is this? A comment like what's done above would help understand.

Reply:
Ah, sure. I'll add a comment.